### PR TITLE
WIP dualstack support for port_cache

### PR DESCRIPTION
**Pull request by Billy99**
Update to port_cache to support dual stack.

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

### CODE REVIEW

**Comment by Billy99** [comment]
NOTE: As Dan Winship pointed out, this conflicts with #1097.

**Comment by Billy99** [comment]
@danwinship PTAL

File: go-controller/pkg/ovn/namespace.go [hunk 1]
```diff
@@ -44,18 +44,22 @@ func (oc *Controller) addPodToNamespace(ns string, portInfo *lpInfo) error {
...
-	if nsInfo.addressSet[address] != "" {
-		return nil
-	}
+	applied := false
+	for _, ip := range portInfo.ips {
+		// If pod has already been added, nothing to do.
+		address := ip.String()
+		if nsInfo.addressSet[address] != "" {
+			applied = true
+			continue
```
Comment (danwinship) [inline go-controller/pkg/ovn/namespace.go:53]: If we assume that "if one IP is already added then all of them have been" then you should just `return nil` here like before. If we don't assume that, then you shouldn't skip the `podAddAllowMulticastPolicy` call when *any* IP is already added, you should only skip it if *all* of the IPs are already added.

File: go-controller/pkg/ovn/common.go [hunk 2]
```diff
@@ -63,8 +63,10 @@ func (oc *Controller) forEachAddressSetUnhashedName(iteratorFn func(
 func addToAddressSet(hashName string, address string) {
 	klog.V(5).Infof("addToAddressSet for %s with %s", hashName, address)
 
+	// IPv6 addresses need to  be quoted, IPv4 work either way.
+	addressQuotes := strconv.Quote(address)
 	_, stderr, err := util.RunOVNNbctl("add", "address_set",
-		hashName, "addresses", address)
+		hashName, "addresses", addressQuotes)
```
Comment (danwinship) [inline go-controller/pkg/ovn/common.go:69]: can you split the quoting stuff (including the corresponding unit test updates) into its own commit, then do the dual-stack stuff as a second commit?

also, you don't really need the temporary variable... you could just change `address` to `strconv.Quote(address)` in the call. Also, you don't *really* need to use `strconv.Quote`... we know the address doesn't contain any quotes or anything else that needs to be escaped, so you could just do `` `"` + address + `"` ``?

Comment (Billy99) [reply go-controller/pkg/ovn/common.go:69]: Will do. I'll move it to another PR so it is easier to manage. Should be posted within the hour.

File: go-controller/pkg/ovn/namespace.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Billy99) [reply go-controller/pkg/ovn/namespace.go]: Fixed. This is only called from 'addLogicalPort()', so assume "if one IP is already added then all of them have been".

**Comment by Billy99** [comment]
@danwinship PTAL

**Comment by danwinship** [comment]
@Billy99 the description in #1142 talks about having separate IPv4 and IPv6 address sets because the docs say:

> An address set may contain Ethernet, IPv4, or IPv6 addresses with optional bitwise or CIDR masks.... A single address set must contain addresses of the same type.

So you've confirmed that it doesn't actually enforce that when *creating* address sets, but does this work for *matching* address sets? If I have an address set "foo" containing both `10.128.0.5` and `fd02::5`, then does it actually work that `ip4.src == $foo` matches traffic from `10.128.0.5` (and nothing else) and `ip6.src == $foo` matches traffic from `fd02::5` (and nothing else)?

**Comment by danwinship** [comment]
elsewhere, @Billy99 said:

> With regards to address_set and supporting multiple IP families simultaneously on a given address_set. I brought up IPv4 deployment and used ovn-nbctl lsp-set-addresses commands to modify the associate IP addresses. Added dual-stack, IPv4 only and IPv6 only. Then dumped the OVN SB Database flows for each to see the differences. Looks like for dual stack, only IPv4 address is applied. So as the documentation indicates, only one address family is supported for a given address_set even though multiple can be entered.

so marking this WIP again; we'll need to go with the separate-address-sets route

**Comment by Billy99** [comment]
Thanks @danwinship, I meant to do that. Sorry about that!

**Comment by Billy99** [comment]
Closing. Content covered under #1373.

**Pull request closed without merging**